scheduler/src/cook/kubernetes/api.clj

```diff
@@ -0,0 +1,316 @@
+  [^V1Pod pod]
+  (when pod
+    (let [^V1PodStatus pod-status (.getStatus pod)
+          container-statuses (.getContainerStatuses pod-status)
+          ; TODO: We want the generation of the pod status to reflect the container status:
+          ; Right now, we only look at one container, the required-cook-job-container container in a pod.
+          ; * In the future, we want support for cook sidecars. In order to determine the pod status
+          ;   from the main cook job status and the sidecar statuses, we'll use a naming scheme for sidecars
+          ; * A pod will be considered complete if all containers with name required-* are complete.
+          ; * The main job container will always be named required-cookJobContainer
+          ; * We want a pod to be considered failed if any container with the name required-* fails or any container
+          ;   with the name extra-* fails.
+          ; * A job may have additional containers with the name aux-*
+          job-status (first (filter (fn [c] (= cook-container-name-for-job (.getName c)))
+                                    container-statuses))]
+      (if (some-> pod .getMetadata .getDeletionTimestamp)
+        ; If a pod has been ordered deleted, treat it as if it was gone, Its being async removed.
+        {:state :missing :reason "Pod was explicitly deleted"}
+        ; If pod isn't being async removed, then look at the containers inside it....
+        (if job-status
+          (let [^V1ContainerState state (.getState job-status)]
+            (cond
+              (.getWaiting state)
+              {:state :pod/waiting
+               :reason (-> state .getWaiting .getReason)}
+              (.getRunning state)
+              {:state :pod/running
+               :reason "Running"}
+              (.getTerminated state)
+              (let [exit-code (-> state .getTerminated .getExitCode)]
+                (if (= 0 exit-code)
+                  {:state :pod/succeeded
+                   :exit exit-code
+                   :reason (-> state .getTerminated .getReason)}
+                  {:state :pod/failed
+                   :exit exit-code
+                   :reason (-> state .getTerminated .getReason)}))
+              :default
+              {:state :pod/unknown
+               :reason "Unknown"}))
+
+          {:state :pod/waiting
+           :reason "Pending"})))))
+
+(defn kill-task
+  "Kill this kubernetes pod"
+  [^ApiClient api-client ^V1Pod pod]
+  (let [api (CoreV1Api. api-client)
+        ^V1DeleteOptions deleteOptions (-> (V1DeleteOptionsBuilder.) (.withPropagationPolicy "Background") .build)]
+    (try
+      (.deleteNamespacedPod
+        api
+        (-> pod .getMetadata .getName)
+        (-> pod .getMetadata .getNamespace)
+        deleteOptions
+        nil
+        nil
+        nil
+        nil
+        nil)
+      (catch JsonSyntaxException e
+        (log/info "Caught the https://github.com/kubernetes-client/java/issues/252 exception.")
```

Comment:
Is it worth also logging the pod namespace and name?

Reply:
No point. We know we deleted it (by the state machine logic that triggers deletion.)